File: customize.ts
#!/usr/bin/env -S deno run --allow-run --allow-read --allow-write
import { Command, ValidationError } from "@cliffy/command";
import { paramCase, pascalCase, snakeCase, titleCase } from "@wok/case";
import { Confirm, Input } from "@cliffy/prompt";
import { Karacho } from "https://deno.land/x/karacho@v1.0.25/main.ts";

await new Command()
  .name("customize.ts")
  .version("1.0.0")
  .description("Customizes the Rust template")
  .arguments("<PROJECTNAME:string>")
  .error((error) => {
    if (error instanceof ValidationError) {
      console.error(error);
      Deno.exit(1);
    }
  })
  .action(async (_options, args) => {
    const oldProjectNameSnake = "rust_cli_quickstart";
    const oldProjectNamePascal = "RustCliQuickStart";
    const oldProjectNameParam = "rust-cli-quickstart";
    const projectName = args;
    const projectNameSnake = snakeCase(projectName);
    const projectNamePascal = pascalCase(projectName);
    const projectNameTitle = titleCase(projectName);
    const projectNameParam = paramCase(projectName);
    const oldBinPath = "src/bin/rust_cli_quickstart.rs";
    const binPath = `src/bin/${projectNameSnake}.rs`;
    const libPath = "src/lib.rs";
    const benchPath = "benches/benchmarks.rs";
    const cargoTomlPath = "Cargo.toml";
    const debugPath = ".zed/debug.json";
    const readMeTemplatePath = "README.template.md";
    const readMePath = "README.md";

    const description = await Input.prompt(
      "Enter the description for the project",
    );
    const firstName = await Input.prompt("Enter your first name");
    const lastName = await Input.prompt("Enter your last name");
    const email = await Input.prompt("Enter your email");
    const alias = await Input.prompt("Enter your GitHub alias");
    const karacho = new Karacho();

    Deno.renameSync(oldBinPath, binPath);

    Deno.writeTextFileSync(
      binPath,
      Deno.readTextFileSync(binPath)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
        .replaceAll(oldProjectNamePascal, projectNamePascal),
    );

    Deno.writeTextFileSync(
      libPath,
      Deno.readTextFileSync(libPath).replaceAll(
        oldProjectNamePascal,
        projectNamePascal,
      ),
    );

    Deno.writeTextFileSync(
      benchPath,
      Deno.readTextFileSync(benchPath)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
        .replaceAll(oldProjectNamePascal, projectNamePascal),
    );

    Deno.writeTextFileSync(
      debugPath,
      Deno.readTextFileSync(debugPath)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
        .replaceAll(oldProjectNameParam, projectNameParam),
    );

    Deno.writeTextFileSync(
      cargoTomlPath,
      karacho.compile(
        Deno.readTextFileSync(cargoTomlPath)
          .replaceAll(oldProjectNameSnake, projectNameSnake)
          .replaceAll(oldProjectNameParam, projectNameParam)
          .replaceAll(oldProjectNamePascal, projectNamePascal),
      )({ description, firstName, lastName, email, alias }),
    );

    Deno.removeSync(readMePath);
    Deno.renameSync(readMeTemplatePath, readMePath);
    Deno.writeTextFileSync(
      readMePath,
      karacho.compile(Deno.readTextFileSync(readMePath))({
        title: projectNameTitle,
        alias,
        projectName: projectNameSnake,
        description,
      }),
    );

    if (await Confirm.prompt("Delete customization scripts?")) {
      Deno.removeSync("customize.ts");
      Deno.removeSync("deno.lock");
      Deno.removeSync("deno.json");
    }

    await new Deno.Command("cargo", { args: ["build"] }).spawn().status;

    if (await Confirm.prompt("Re-initialize Git repo?")) {
      Deno.removeSync(".git", { recursive: true });
      await new Deno.Command("git", { args: ["init"] }).spawn().status;
      await new Deno.Command("git", { args: ["add", "-A", ":/"] }).spawn()
        .status;
      await new Deno.Command("git", {
        args: ["commit", "-m", "Initial commit"],
      }).spawn().status;
    }

    console.info("Customization complete");
  })
  .parse(Deno.args);

File: customize.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! yansi = "0.5.1"
//! str_inflector = "0.12.0"
//! dialoguer = "0.10.3"
//! handlebars = "4.3.6"
//! map-macro = "0.2.5"
//! duct = "0.13.6"
//! ```

use dialoguer::{Confirm, Input};
use duct::cmd;
use handlebars::Handlebars;
use inflector::Inflector;
use map_macro::map;
use std::{env, error::Error, fs, process::ExitCode};
use yansi::Paint;

trait Logger {
    fn info<T: AsRef<str>>(s: T);
    fn error<T: AsRef<str>>(s: T);
    fn warning<T: AsRef<str>>(s: T);
}

struct Customizer {}

impl Logger for Customizer {
    fn info<T: AsRef<str>>(s: T) {
        println!("👉 {}", s.as_ref());
    }

    fn error<T: AsRef<str>>(s: T) {
        eprintln!("💥 {}", Paint::red(s.as_ref()));
    }

    fn warning<T: AsRef<str>>(s: T) {
        eprintln!("🐓 {}", Paint::yellow(s.as_ref()));
    }
}

impl Customizer {
    fn new() -> Customizer {
        Customizer {}
    }

    fn run(&self, project_name: &str) -> Result<(), Box<dyn Error>> {
        let project_name_snake = project_name.to_snake_case();
        let project_name_pascal = project_name.to_pascal_case();
        let project_name_title = project_name.to_title_case();
        let project_name_kebab = project_name.to_kebab_case();
        let old_bin_path = "src/bin/rust_cli_quickstart.rs";
        let bin_path = format!("src/bin/{}.rs", &project_name.to_snake_case());
        let lib_path = "src/lib.rs";
        let bench_path = "benches/benchmarks.rs";
        let cargo_toml_path = "Cargo.toml";
        let launch_path = ".vscode/launch.json";
        let read_me_path = "README.md";

        fs::rename(old_bin_path, &bin_path)?;

        fs::write(
            &bin_path,
            fs::read_to_string(&bin_path)?
                .replace("rust_cli_quickstart", &project_name_snake)
                .replace("RustCliQuickStart", &project_name_pascal),
        )?;

        fs::write(
            &lib_path,
            fs::read_to_string(&lib_path)?.replace("RustCliQuickStart", &project_name_pascal),
        )?;

        fs::write(
            &bench_path,
            fs::read_to_string(&bench_path)?
                .replace("rust_cli_quickstart", &project_name_snake)
                .replace("RustCliQuickStart", &project_name_pascal),
        )?;

        fs::write(
            &cargo_toml_path,
            fs::read_to_string(&cargo_toml_path)?
                .replace("rust_cli_quickstart", &project_name_snake)
                .replace("rust-cli-quickstart", &project_name_kebab)
                .replace("RustCliQuickStart", &project_name_pascal),
        )?;

        fs::write(
            &launch_path,
            fs::read_to_string(&launch_path)?
                .replace("rust_cli_quickstart", &project_name.to_snake_case())
                .replace("RustCliQuickStart", &project_name.to_pascal_case()),
        )?;

        let description = Input::<String>::new()
            .with_prompt("Enter the description for the project")
            .interact_text()?;
        let first_name = Input::<String>::new()
            .with_prompt("Enter your first name")
            .interact_text()?;
        let last_name = Input::<String>::new()
            .with_prompt("Enter your last name")
            .interact_text()?;
        let email = Input::<String>::new()
            .with_prompt("Enter your email")
            .interact_text()?;
        let alias = Input::<String>::new()
            .with_prompt("Enter your GitHub alias")
            .interact_text()?;

        let handlebars = Handlebars::new();

        fs::write(
            &cargo_toml_path,
            &handlebars.render_template(
                &fs::read_to_string(&cargo_toml_path)?,
                &map! {
                    "description" => &description,
                    "firstName" => &first_name,
                    "lastName" => &last_name,
                    "email" => &email,
                    "alias" => &alias,
                },
            )?,
        )?;

        fs::write(
            &read_me_path,
            &handlebars.render_template(
                &fs::read_to_string(&read_me_path)?,
                &map! {
                    "title" => &project_name_title,
                    "alias" => &alias,
                    "projectName" => &project_name_snake,
                    "description" => &description,
                },
            )?,
        )?;

        if Confirm::new()
            .with_prompt("Delete customization scripts?")
            .interact()?
        {
            fs::remove_file("customize.rs")?;
        }

        if Confirm::new()
            .with_prompt("Reinitialize Git repo?")
            .interact()?
        {
            fs::remove_dir_all(".git")?;
            cmd!("git", "init").run()?;
            cmd!("git", "add", "-A", ":/").run()?;
            cmd!("git", "commit", "-m", "Initial commit").run()?;
        }

        Self::info("Customization complete");

        Ok(())
    }
}

fn main() -> ExitCode {
    let args: Vec<String> = env::args().collect();
    let customizer = Customizer::new();

    if args.len() < 2 {
        eprintln!("Usage: customize [PROJECT-NAME]");
        return ExitCode::SUCCESS;
    }

    if let Err(err) = customizer.run(&args[1]) {
        Customizer::error(&err.to_string());
        return ExitCode::FAILURE;
    }

    ExitCode::SUCCESS
}
